### PR TITLE
feat(streams): reactive primitives for phlex-reactive (#165–#171)

### LINKED ISSUE #165: Streams: per-connection actor-echo suppression (exclude: on broadcast)
## Motivation (from phlex-reactive)

When a user triggers a reactive action that also broadcasts (e.g. "send message"), they receive **both** the action's HTTP response AND the SSE echo of their own broadcast. Today consumers lean on idiomorph DOM-id dedup to swallow the duplicate. That works for `append`/`replace` of an element with a stable id, but is fragile for ordered inserts, optimistic UI, animations (the echo re-runs them), and non-idempotent actions.

## What pgbus already has

`Pgbus::Streams::Filters` — per-connection, server-side audience predicates evaluated against each connection's authorize-context (`visible_to: :label`). **The delivery-time, per-connection filtering machinery already exists.**

## Proposal

Give every SSE connection a stable **connection id** (mint it server-side, expose to the page via a `<meta name="pgbus-connection-id">` or on `<pgbus-stream-source>`). Let a broadcast carry `exclude:` / `origin:` connection id; the dispatcher skips delivery to that connection — reusing the Filters delivery path.

```ruby
Pgbus.stream("chat", room).broadcast(html, exclude: request.headers["X-Pgbus-Connection"])
# and via Turbo::StreamsChannel.broadcast_*_to(..., exclude:)
```

## Why it's #1

Removes the sharpest correctness edge in server-driven UIs (double-apply), makes optimistic UI safe, and enables the clean Livewire model: "the actor's truth is the HTTP response; everyone else gets the broadcast." The hook (`exclude` → Filters-style skip) is small given what exists.

### PR DESCRIPTION
**Pull request by mhenrixon**
## Summary

Implements the seven Streams primitives that **phlex-reactive** needs to build truly reactive Rails websites on pgbus's SSE pub/sub. Each is a small, opt-in, backward-compatible addition to the existing Streams subsystem, shipped TDD as one focused commit.

Closes #165, #166, #167, #168, #169, #170, #171.

## The primitives

| # | Primitive | What it does | phlex-reactive call-site it unblocks |
|---|-----------|--------------|--------------------------------------|
| #167 | `stream_key` idempotent | A single `String` key passes through unchanged (after the budget check) instead of tripping the colon guard; `Pgbus.stream_key!` for the explicit pre-built case | Pass one `stream_key` value to both `turbo_stream_from` and the broadcaster without `ArgumentError` |
| #165 | actor-echo `exclude:` | `broadcast(html, exclude: connection_id)` skips delivery to the actor's own SSE connection (before audience filtering). The server mints + exposes a per-connection id via a new `pgbus:connected` frame | The clean Livewire model: the actor's truth is the action's HTTP response; everyone else gets the broadcast — no double-apply |
| #166 | `broadcast_render` | `broadcast_render(renderable:, target:, action:)` renders a Phlex component / ViewComponent / string into a `<turbo-stream>` tag and broadcasts atomically | `broadcast_*_to` becomes a thin shim; the off-request view context is built correctly in one place |
| #168 | msg_id → client | Each frame's msg_id reaches the client as the `message` event's `lastEventId` and a new `pgbus:message { msgId, data }` event; documents the skip-if-newer-rev pattern | Safe optimistic UI via a per-target `maxAppliedRev` map, no bespoke versioning |
| #169 | auto-presence | `streams_presence_patterns` + `streams_presence_member` drive auto join/leave/touch from the connection lifecycle | A `<Presence>` reactive component becomes a one-flag feature |
| #170 | typed SSE events | `broadcast(html, event: "presence")` sets the SSE `event:` while keeping a Turbo Stream payload; the client routes via `pgbus:event` / `pgbus:<name>` | Route presence pills, analytics, and sound-on-new-message without sniffing HTML |
| #171 | coalescing | `broadcast(coalesce:, target:)` keeps only the latest frame per (stream, target) within a window — superseded frames never hit the bus | High-frequency cursors / typing / progress without flooding the bus |

## Composition

The primitives compose: `broadcast_render(..., exclude:, visible_to:, event:, coalesce:)` threads all of them through one call. `exclude` is evaluated before the audience filter; coalescing re-enters the normal broadcast path so a coalesced frame is just a deferred ordinary broadcast.

## Wire-format additions (all opt-in, backward-compatible)

- **JSONB payload** gains `exclude` and `event` fields alongside the existing `html` / `visible_to`.
- **New SSE frame** `event: pgbus:connected` carries the server-minted connection id, written once after the open handshake (Puma hijack + Falcon streaming-body paths).
- **`StreamEnvelope`** gains `exclude` and `event` fields.
- **New config**: `streams_presence_patterns`, `streams_presence_member` (mirrors `streams_durable_patterns`).
- **New client surface** on `<pgbus-stream-source>`: events `pgbus:connected`, `pgbus:message`, `pgbus:event`, `pgbus:<name>`; attributes `connection-id`, `listen-events`.

## Architecture notes

- pgbus keeps **no hard dependency** on turbo-rails / ActionView / Phlex — `broadcast_render`'s `Renderer` is self-contained and resolves a renderable via `String` → `#call` (Phlex) → `#render_in` (ViewComponent) → `#to_s`.
- Presence membership work runs on the **dispatcher thread** (which already releases AR connections each pass); the heartbeat posts a batched `PresenceTouchMessage` per tick. Presence failures are logged and swallowed so a presence-DB hiccup can't drop a live SSE connection.
- The `Coalescer` is a process-wide, thread-safe, trailing-edge-with-max-wait debounce; the scheduler is injectable (`Concurrent::ScheduledTask` in prod, a synchronous fake in unit tests).

## Test plan

- **Unit specs** (TDD, RED→GREEN per commit) for every changed file: `stream_key`, `Renderer`, `Coalescer`, `Stream#broadcast`/`#broadcast_render`, dispatcher (`exclude`, presence), connection/falcon (typed events, presence_member), heartbeat (presence touch), envelope (`connected`), configuration (presence patterns/member). All green; rubocop clean.
- **Integration specs** (real Postgres + PGMQ + Puma SSE harness): `exclude_spec`, `auto_presence_spec`, `typed_event_spec` added; existing `audience` / `replay` / `reconnect` / `presence` / `falcon` specs unperturbed (the `pgbus:connected` control frame is routed separately in `SseTestClient` so delivery assertions don't shift). **41 streams integration examples, 0 failures.**
- Client JS verified to parse and the `msgId` parsing edge cases (numeric / negative-ephemeral / zero / null) checked with `node`.

> Pre-existing failures in `spec/system/*` (browser/Capybara dashboard tests) and `spec/i18n_spec.rb` are unrelated to this PR — they fail identically with these changes stashed, and none of the changed files touch system tests, i18n, views, or controllers.

## Docs

`CHANGELOG.md` updated with one entry per primitive under `[Unreleased] > Added`.

## Release Notes

* **New Features**
  * Connection-driven presence tracking for SSE streams (auto-join/leave plus heartbeat “touch”).
  * Opt-in high-frequency broadcast coalescing with `coalesce`, `target`, and `window_ms`.
  * Typed SSE `event:` names with safer typed delivery end-to-end.
  * Actor-echo suppression via `exclude`, using a per-connection id now exposed to clients.
  * Stream `msg_id`/revision surfaced to support optimistic UI updates.
  * Enhanced `broadcast_render` to render and broadcast with `event` and `coalesce`.
  * `Pgbus.stream_key!` to accept pre-built stream keys.
* **Bug Fixes**
  * Hardened SSE frame encoding to prevent event/data injection via newline characters.

### CODE REVIEW

File: app/assets/javascripts/pgbus/stream_source_element.js [hunk 1]
```diff
@@ -99,7 +127,7 @@ class PgbusStreamSourceElement extends HTMLElement {
 
       this.setAttribute("connected", "")
       this.dispatchEvent(new CustomEvent("pgbus:open", {
-        detail: { lastEventId: this.lastEventId }
+        detail: { lastEventId: this.lastEventId, connectionId: this.connectionId }
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply app/assets/javascripts/pgbus/stream_source_element.js:136]: Fixed in c881f1a. `<pgbus-stream-source>` now calls `resetConnectionId()` (clearing `this.connectionId` and the `connection-id` attribute) at the start of both `openFetchStream()` and `switchToEventSource()`, so `pgbus:open` can't surface the previous connection's id during a reconnect window.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,21 @@
...
+
+- **Streams: typed SSE event names on broadcasts.** A broadcast can set the SSE `event:` field — `Pgbus.stream(name).broadcast(html, event: "presence")` (and `broadcast_render(..., event:)`) — while keeping the payload a Turbo Stream, so clients route on the typed name instead of sniffing the HTML. The default (`nil` or `"turbo-stream"`) is unchanged and not carried on the wire; default consumers still get the standard `message`/turbo-stream path. The event flows through the JSONB payload → `StreamEnvelope.event` → the SSE frame's `event:` line (both the Puma and Falcon connection adapters). On the client, `<pgbus-stream-source>` dispatches a typed broadcast as a generic `pgbus:event` (`{ event, data, msgId }`) and a named `pgbus:<event>` (`{ data, msgId }`) for `addEventListener` ergonomics; declare typed event names via the element's `listen-events` attribute so they survive the EventSource reconnect path. Refs #170.
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply CHANGELOG.md]: Fixed in c881f1a. Reworded: the default event is omitted from the JSONB payload to avoid redundancy but is still set on the SSE frame's `event:` line (the connection adapter falls back to `turbo-stream` via `sse_event_for`).

File: lib/pgbus/configuration.rb [hunk 3]
```diff
@@ -411,6 +421,8 @@ def validate_streams!
 
       raise ArgumentError, "streams_durable_patterns must be an Array of strings/regex" unless streams_durable_patterns.is_a?(Array)
 
+      raise ArgumentError, "streams_presence_patterns must be an Array of strings/regex" unless streams_presence_patterns.is_a?(Array)
+
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/configuration.rb:425]: Fixed in c881f1a. `Configuration#validate!` now raises `ArgumentError` unless `streams_presence_member` is nil or responds to `#call`, so a non-callable fails fast at boot instead of at runtime in `presence_member_for`.

File: lib/pgbus/streams.rb [hunk 4]
```diff
@@ -76,9 +101,48 @@ def durable?
...
+          return coalesce_broadcast(
+            payload,
+            coalesce: coalesce,
+            target: target,
+            visible_to: visible_to,
+            durable: durable,
+            exclude: exclude,
+            event: event
+          )
+        end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/streams.rb:140]: Fixed in c881f1a. A coalesced *durable* broadcast made inside an open transaction now defers the coalescer submission to `after_commit` (mirrors the non-coalesced path), so a rollback no longer flushes the frame. Ephemeral coalescing remains fire-and-forget.

File: lib/pgbus/streams.rb [hunk 5]
```diff
@@ -172,6 +266,30 @@ def broadcast_ephemeral(wrapped)
...
+      def coalesce_broadcast(payload, coalesce:, target:, visible_to:, durable:, exclude:, event:)
+        if target.nil? || target.to_s.empty?
+          raise ArgumentError,
+                "broadcast(coalesce:) requires target: — coalescing keys on (stream, target), " \
+                "so the latest frame per target can win. Use broadcast_render(coalesce:) which " \
+                "already has a target, or pass target: explicitly."
+        end
+
+        window_ms = coalesce == true ? Coalescer::DEFAULT_WINDOW_MS : coalesce
+        Streams.coalescer.submit(
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/streams.rb]: Fixed in c881f1a. `broadcast(coalesce:)` now validates at the API boundary: `true` → default window, a positive Numeric is used as-is, anything else raises an actionable `ArgumentError` instead of a later `NoMethodError` in the timer math.

File: lib/pgbus/streams.rb [hunk 6]
```diff
@@ -172,6 +266,30 @@ def broadcast_ephemeral(wrapped)
...
+          stream_name: @name,
+          target: target.to_s,
+          payload: payload.to_s,
+          window_ms: window_ms,
+          opts: { visible_to: visible_to, durable: durable, exclude: exclude, event: event }
+        )
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/streams.rb]: Fixed in c881f1a. `coalesce_broadcast` now resolves `durable: nil` against the originating stream instance (`@durable`) before submitting, so the coalescer flush re-entering `Pgbus.stream(name)` can't change the durability mode for an explicitly-ephemeral stream.

File: lib/pgbus/web/streamer/stream_event_dispatcher.rb [hunk 7]
```diff
@@ -239,14 +261,17 @@ def handle_ephemeral_wake(msg, stream, registered, in_flight_pairs, started_at)
...
           envelope = StreamEnvelope.new(
             msg_id: -@ephemeral_seq,
             enqueued_at: Time.now.utc.iso8601(6),
             payload: html,
             source: "ephemeral",
-            visible_to: visible_to
+            visible_to: visible_to,
+            exclude: exclude,
+            event: parsed["event"]
           )
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/web/streamer/stream_event_dispatcher.rb:275]: Fixed in c881f1a. Defense in depth: `Envelope.message` now strips CR/LF from `event` as well as `data`, and the dispatcher's new `normalize_sse_event` drops non-strings, blanks, and any value containing CR/LF (→ default turbo-stream) before building the `StreamEnvelope` on both the ephemeral and unwrap paths.

File: spec/integration/streams/auto_presence_spec.rb [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    Pgbus.reset_client!
+  end
+
+  after(:all) do
+    Pgbus.configuration.listen_notify = @saved_listen_notify
+    Pgbus.configuration.streams_presence_patterns = @saved_presence_patterns
+    Pgbus.configuration.streams_heartbeat_interval = @saved_heartbeat
+    Pgbus.configuration.streams_signed_name_secret = nil
+    Pgbus.reset_client!
+  end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/integration/streams/auto_presence_spec.rb:45]: Fixed in c881f1a. `before(:all)` now saves `streams_signed_name_secret`, `streams_listen_health_check_ms`, and `streams_write_deadline_ms`, and `after(:all)` restores all mutated globals to their saved values instead of hard-setting the secret to nil.

File: spec/integration/streams/exclude_spec.rb [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    Pgbus.configuration.streams_heartbeat_interval = 30
+    Pgbus.configuration.streams_write_deadline_ms = 5_000
+    Pgbus.reset_client!
+  end
+
+  after(:all) do
+    Pgbus.configuration.listen_notify = @saved_listen_notify
+    Pgbus.configuration.streams_signed_name_secret = nil
+    Pgbus.reset_client!
+  end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/integration/streams/exclude_spec.rb:42]: Fixed in c881f1a. `before(:all)` now saves all five mutated globals and `after(:all)` restores each to its prior value (no more hard-set to nil).

File: spec/integration/streams/typed_event_spec.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    Pgbus.configuration.streams_heartbeat_interval = 30
+    Pgbus.configuration.streams_write_deadline_ms = 5_000
+    Pgbus.reset_client!
+  end
+
+  after(:all) do
+    Pgbus.configuration.listen_notify = @saved_listen_notify
+    Pgbus.configuration.streams_signed_name_secret = nil
+    Pgbus.reset_client!
+  end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/integration/streams/typed_event_spec.rb:32]: Fixed in c881f1a. `before(:all)` now saves `streams_signed_name_secret`, `streams_listen_health_check_ms`, `streams_heartbeat_interval`, and `streams_write_deadline_ms`, and `after(:all)` restores each to its saved value.
